src/arch/arm/isa/insts/sve.isa

```diff
@@ -2096,6 +2096,34 @@ let {{
                          'class_name' : 'Sve' + Name}
             exec_output += SveOpExecDeclare.subst(substDict)
 
+    # Generates definitions for ternary SVE instructions (indexed)
+    def sveTerInstIndexed(name, Name, opClass, types, op, decoder='Generic'):
+        global header_output, exec_output, decoders
+        code = sveEnabledCheckCode + '''
+        unsigned eCount = ArmStaticInst::getCurSveVecLen<Element>(
+                xc->tcBase());
+        for (unsigned i = 0; i < eCount; i++) {
+            int segbase = i - i % (128 / sizeof(Element));
+            int s = segbase + imm;
+            const Element& srcElem1 = AA64FpOp1_x[i];
+            const Element& srcElem2 = AA64FpOp2_x[s];
+            Element destElem = AA64FpDestMerge_x[i];
+            '''
+
+        code +=  f"{op} \n"
+
+
+        code += ''' AA64FpDest_x[i] = destElem;
+        }'''
+        iop = ArmInstObjParams(name, 'Sve2' + Name+ 'i', 'SveTerIndexedOp',
+                               {'code': code, 'op_class': opClass}, [])
+        header_output += SveTerIndexedOpDeclare.subst(iop)
+        exec_output += SveOpExecute.subst(iop)
+        for type in types:
+            substDict = {'targs' : type,
+                         'class_name' : 'Sve2' + Name + "i"}
+            exec_output += SveOpExecDeclare.subst(substDict)
+
```

Comment:
This is effectively the same as sveBinIdxInst.. I believe we should reuse existing code

Reply:
Hi,

Thank you for your suggestions, while I make the changes I have a few observations I would like to bring to you.
the sveBinIdxInst function, here would might not work as it does the following 

```
Element destElem = 0;
```
Instead of 
```
Element destElem = AA64FpDestMerge_x[i];
```

which is needed for mla/s indexed. However, as you mentioned the template library does have some repeated code and I see a  ```sveTerIdxInst``` function that existed before and solves the purpose I shall use that.


Also, I have another observation and would like to know more about why the implementation is done this way. For both ```sveTerIdxInst``` and ```sveBinIdxInst``` there is a temporary  ```VegRegContainer``` created as ```auxDest``` ( seems like auxiliary destination), which then is written into ``` AA64FpDest_x[i] ``` using a separate for loop as below:

```
    Element destElem = AA64FpDestMerge_x[i];
        '''
        code += '''
            %(op)s
            auxDest[i] = destElem;
        }

        for (unsigned i = 0; i < eCount; i++) {
            AA64FpDest_x[i] = auxDest[i];
        }''' % {'op': op}
```

Is there a reason why this is done instead of using ```AA64FpDest_x[i]``` in the previous for-loop (like other instructions that use the ```sveTerInst``` function?

Let me know.

Regards

Reply:
> Hi,
> 
> Thank you for your suggestions, while I make the changes I have a few observations I would like to bring to you. the sveBinIdxInst function, here would might not work as it does the following
> 
> ```
> Element destElem = 0;
> ```
> 
> Instead of
> 
> ```
> Element destElem = AA64FpDestMerge_x[i];
> ```
> 

Oh right I missed that!

> which is needed for mla/s indexed. However, as you mentioned the template library does have some repeated code and I see a `sveTerIdxInst` function that existed before and solves the purpose I shall use that.
> 

That's great!

> Also, I have another observation and would like to know more about why the implementation is done this way. For both `sveTerIdxInst` and `sveBinIdxInst` there is a temporary `VegRegContainer` created as `auxDest` ( seems like auxiliary destination), which then is written into `AA64FpDest_x[i]` using a separate for loop as below:
> 
> ```
>     Element destElem = AA64FpDestMerge_x[i];
>         '''
>         code += '''
>             %(op)s
>             auxDest[i] = destElem;
>         }
> 
>         for (unsigned i = 0; i < eCount; i++) {
>             AA64FpDest_x[i] = auxDest[i];
>         }''' % {'op': op}
> ```
> 
> Is there a reason why this is done instead of using `AA64FpDest_x[i]` in the previous for-loop (like other instructions that use the `sveTerInst` function?
> 
> Let me know.
> 
> Regards

This could potentially be an unnecessary copy. I have two possible explanations:

1) To handle the case where the source = destination register, meaning

`AA64FpDest_x[i] = destElem`

might change the source register value while still iterating. This is not the case anyway, as while writing to the destination register changes the physical register, the source operands are copied anyway into temporaries tmp_s* with getRegOperand

2) To handle FP exceptions thrown while iterating over the vectors. In case of an exception we don't want to have an inconsistent register state.. FP exceptions are not implemented in Arm anyway so again, there seem to be non need for that.

I will run some tests and see if we can safely remove that auxiliary vector.. Anyway it should be indipendent from this PR. Thanks for flagging it though!

Reply:
Thank you for clearing the doubts, do let me know if I can do anything regarding this. I would like to be involved with it.